### ISSUE TITLE
Streaming mdoe for schema generation

### ISSUE THREAD

**Issue by fgregg** [wireservice/csvkit#913]
Related to conversation in #737, i'd really like a memory efficient (and faster) version of schema generation. Let's start by describing how csvkit currently generates a schema (this is a bit simplified)

1. `csvsql` passes a file object to the [`from_csv` class method of Table](https://github.com/wireservice/csvkit/blob/e10d4a3caf5a88b5577bd2e8b69419c4024d3e9b/csvkit/utilities/csvsql.py#L124-L130)
2. `from_csv` reads the file as a csv an [saves all rows to a `rows` object](https://github.com/wireservice/csvkit/blob/e10d4a3caf5a88b5577bd2e8b69419c4024d3e9b/csvkit/utilities/csvsql.py#L124-L130)
3. `from_csv` passes `rows` to `Table.__init__`
3. `Table __init__` iterates through all rows to do [type inference](https://github.com/wireservice/agate/blob/master/agate/table/__init__.py#L109)
4. `Table __init__` iterates through all rows and [casts fields to appropriate type and saving result to a a `new_rows`](https://github.com/wireservice/agate/blob/master/agate/table/__init__.py#L116-L135). This effectively doubles the memory footprint until we exit from `__init__` and `rows` is garbage collected
5.  the `Table __init__` method exits and control is passed to the `csvsql` script
6.  `csvsql` calls `to_sql_create_statement` method is called on the Table object
7. `to_sql_create_statement` calls the `make_sql_table` method
8. if user is not using the `--no-constraints` flag on `csvsql` the `make_sql_table` [calculates precision and length constraints for every column](https://github.com/wireservice/agate-sql/blob/2be303738bbabea5471b0f6844763fd559571cf4/agatesql/table.py#L171-L187). This effectively doubles the memory footprint because the agate aggregation methods make calls `column.values` and `column.values_without_nulls` which create list objects of the values in a column. This methods are memoized, so these lists will remain in memory until the parent column object is garbage collected, which won't happen until csvsql exits.
9. `<agate-sql>` The column information is passed to `make_sql_column` which makes the appropriate sql schema entry for that column.

Here's an alternative flow that I would suggest for a streaming process:

1. csv reader is passed to a type inference that is similar to existing type inference. Exhaust all the rows. In addition to doing the type inference, we will also get information about max length and precision in this iteration
2. this information is passed to `make_sql_column' 

So, basically, bypassing creation of the Table object and expanding the functionality of the type tester.

Thoughts?

**Comment by jpmckinney** [comment]
This seems reasonable to me!
